src/sources/audiosourcemp3.cpp

```diff
@@ -0,0 +1,531 @@
+
+const AudioSource::sample_type kMadScale = AudioSource::kSampleValuePeak
+        / AudioSource::sample_type(MAD_F_ONE);
+
+inline AudioSource::sample_type madScale(mad_fixed_t sample) {
+    return sample * kMadScale;
+
+} // anonymous namespace
+
+// Optimization: Reserve initial capacity for seek frame list
+const AudioSource::size_type kMinutesPerFile = 10; // enough for the majority of files (tunable)
+const AudioSource::size_type kSecondsPerMinute = 60; // fixed
+const AudioSource::size_type kMaxMp3FramesPerSecond = 39; // fixed: 1 MP3 frame = 26 ms -> ~ 1000 / 26
+const AudioSource::size_type kSeekFrameListCapacity = kMinutesPerFile
+        * kSecondsPerMinute * kMaxMp3FramesPerSecond;
+
+int decodeFrameHeader(
+        mad_header* pMadHeader,
+        mad_stream* pMadStream,
+        bool skipId3Tag) {
+    const int result = mad_header_decode(pMadHeader, pMadStream);
+    if ((0 != result) && (MAD_ERROR_LOSTSYNC == pMadStream->error)
+            && skipId3Tag) {
+        long tagsize = id3_tag_query(pMadStream->this_frame,
+                pMadStream->bufend - pMadStream->this_frame);
+        if (0 < tagsize) {
+            // Skip ID3 tag data
+            mad_stream_skip(pMadStream, tagsize);
+        }
+    }
+    return result;
+}
+
+} // anonymous namespace
+
+AudioSourceMp3::AudioSourceMp3(QUrl url)
+        : AudioSource(url),
+          m_file(getUrl().toLocalFile()),
+          m_fileSize(0),
+          m_pFileData(NULL),
+          m_avgSeekFrameCount(0),
+          m_curFrameIndex(kFrameIndexMin),
+          m_madSynthCount(0) {
+    mad_stream_init(&m_madStream);
+    mad_stream_options(&m_madStream, MAD_OPTION_IGNORECRC);
+    mad_frame_init(&m_madFrame);
+    mad_synth_init(&m_madSynth);
+    m_seekFrameList.reserve(kSeekFrameListCapacity);
+}
+
+AudioSourceMp3::~AudioSourceMp3() {
+    close();
+}
+
+AudioSourcePointer AudioSourceMp3::create(QUrl url) {
+    QSharedPointer<AudioSourceMp3> pAudioSource(new AudioSourceMp3(url));
+    if (OK == pAudioSource->open()) {
+        // success
+        return pAudioSource;
+    } else {
+        // failure
+        return AudioSourcePointer();
+    }
+}
+
+Result AudioSourceMp3::open() {
+    if (!m_file.open(QIODevice::ReadOnly)) {
+        qWarning() << "Failed to open file:" << m_file.fileName();
+        return ERR;
+    }
+
+    // Get a pointer to the file using memory mapped IO
+    m_fileSize = m_file.size();
+    m_pFileData = m_file.map(0, m_fileSize);
+    // Transfer it to the mad stream-buffer:
+    mad_stream_buffer(&m_madStream, m_pFileData, m_fileSize);
+    DEBUG_ASSERT(m_pFileData == m_madStream.this_frame);
+
+    // Decode all the headers and calculate audio properties
+
+    setChannelCount(kChannelCountDefault);
+    setFrameRate(kFrameRateDefault);
+
+    mad_units madUnits = MAD_UNITS_44100_HZ; // default value
+    mad_timer_t madDuration = mad_timer_zero;
+    unsigned long sumBitrate = 0;
+
+    m_curFrameIndex = kFrameIndexMin;
```

Comment:
some initializations are redundant, because already initialized in the initializer list above.